File: Java/11_BinaryTreeGeneral/LC_314_BinaryTreeVerticalOrderTraversal.java
import java.util.*;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solutio_314 {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        if(root == null){
            return new ArrayList<>();
        }

        Map<Integer,List<Integer>> res = new TreeMap();
        Map<TreeNode,Integer> nodeMap = new HashMap();
        nodeMap.put(root,0);

        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        while(!queue.isEmpty()){
            TreeNode node = queue.poll();
            int i = nodeMap.get(node);
            res.computeIfAbsent(i,k -> new ArrayList<>()).add(node.val);
            if(node.left != null){
                queue.add(node.left);
                nodeMap.put(node.left,i - 1);
            }
            if(node.right != null){
                queue.add(node.right);
                nodeMap.put(node.right,i + 1);
            }
        }

        return new ArrayList<>(res.values());
    }
}

public class LC_314_BinaryTreeVerticalOrderTraversal {

}

File: JavaScript/12_BinaryTree/LC_314_BinaryTreeVerticalOrderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var verticalOrder = function (root) {
    if (!root) {
        return [];
    }

    var queue = [[root, 0]];
    var col_map = new Map();

    while (queue.length) {
        [node, col] = queue.shift();

        if (col_map.has(col)) {
            col_map.get(col).push(node.val);
        }
        else {
            col_map.set(col, [node.val]);
        }

        if (node.left) {
            queue.push([node.left, col - 1]);
        }

        if (node.right) {
            queue.push([node.right, col + 1]);
        }
    }

    return getColRes(col_map);

    // 按列的顺序获取最终结果，如果是有序哈希表，则可以省略这一步，直接输出结果
    function getColRes(map) {
        let keys = Array.from(map.keys());
        keys.sort((a, b) => a - b);

        let res = [];
        keys.forEach(key => {
            res.push(map.get(key))
        })

        return res;
    }
};